node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type CreateProposer = Pin<Box<
+		dyn Future<Output = Result<Self::Proposer, Self::Error>> + Send + 'static
+	>>;
+	type Proposer = Proposer<TxPool, Backend, Client>;
+	type Error = sp_blockchain::Error;
+
+	fn init(&mut self, parent_header: &Header) -> Self::CreateProposer {
+		unimplemented!()
+	}
+}
+
+pub struct Proposer<TxPool: TransactionPool<Block = Block>, Backend, Client> {
+	inner: sc_basic_authorship::Proposer<Backend, Block, Client, TxPool>,
+	client: Arc<Client>,
+	overseer: OverseerHandler,
+}
+
+impl<TxPool, Backend, Client> sp_consensus::Proposer<Block> for Proposer<TxPool, Backend, Client>
+where
+	TxPool: 'static + TransactionPool<Block = Block>,
+	Client: 'static
+		+ BlockBuilderProvider<Backend, Block, Client>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Send
+		+ Sync,
+	Client::Api:
+		ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
+	Backend:
+		'static + sc_client_api::Backend<Block, State = sp_api::StateBackendFor<Client, Block>>,
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type Transaction = sc_client_api::TransactionFor<Backend, Block>;
+	type Proposal = Pin<Box<dyn Future<Output = Result<Proposal<Block, sp_api::TransactionFor<Client, Block>>, Error>> + Send>>;
+	type Error = Error;
+
+	fn propose(
+		self,
+		mut inherent_data: InherentData,
+		inherent_digests: DigestFor<Block>,
+		max_duration: time::Duration,
+		record_proof: RecordProof,
+	) -> Self::Proposal {
+		// REVIEW: per the guide, block authors must re-send a new request for block authorship data
+		// for each block. I'm assuming that this function gets called once per block, but it is not
+		// obvious from the documentation that that assumption is in fact true.
+		let mut bitfields = Vec::new();
+		let mut candidates = Vec::new();
+
+		async move {
+			let (sender, receiver) = futures::channel::mpsc::channel(1);
+			self.overseer.send_msg(AllMessages::Provisioner(ProvisionerMessage::RequestBlockAuthorshipData(
+				// REVIEW: is this in fact the hash we want to use here?
+				self.client.info().best_hash,
+				sender,
+			)));
+			receiver.for_each_concurrent(None, |item| {
+				match item {
+					ProvisionableData::Bitfield(_, signed_bitfield) => bitfields.push(signed_bitfield),
+					ProvisionableData::BackedCandidate(candidate) => candidates.push(candidate),
+					_ => {},
+				}
+			});
+			inherent_data.put_data(inclusion_inherent::INHERENT_IDENTIFIER, &(bitfields, candidates))?;
+			self.inner.propose(inherent_data, inherent_digests, max_duration, record_proof).await.map_err(Into::into)
+		}.boxed()
+	}
+}
+
+#[derive(Debug)]
+pub enum Error {
```

Comment:
Does a general-purpose error enum already exist, or is it fine to just declare my own here?

Reply:
Substrate's `ClientError` is commonly used, but the new error enum seems fine for now.